File: boilerplate/java/order-service/src/main/java/com/example/orderservice/domain/ports/OrderRepository.java
package com.example.orderservice.domain.ports;

import com.example.orderservice.domain.models.Order;
import com.example.orderservice.domain.models.OrderId;
import com.example.orderservice.domain.models.OrderState;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Repository port for Order aggregate.
 *
 * All queries excluding {@link #findByIdIncludingDeleted} filter out
 * soft-deleted orders (deleted_at IS NULL).
 */
public interface OrderRepository {
    Order save(Order order);
    Optional<Order> findById(OrderId id);
    Optional<Order> findByIdIncludingDeleted(OrderId id);
    List<Order> findAll();
    List<Order> findAllIncludingDeleted();

    /**
     * Find paginated orders for a customer, filtering by status and excluding soft-deleted.
     *
     * @param customerId the owning customer
     * @param status     optional status filter (null = all statuses)
     * @param page       0-based page number
     * @param size       page size
     * @return page of orders sorted by createdAt descending
     */
    List<Order> findByCustomerId(UUID customerId, OrderState status, int page, int size);

    long countByCustomerId(UUID customerId, OrderState status);
    void deleteById(OrderId id);
    long count();
    boolean existsById(OrderId id);
}

File: boilerplate/python/order-service/src/domain/ports/order_repository.py
from abc import ABC, abstractmethod
from typing import Optional, List

from domain.order import Order
from domain.order_id import OrderId

class OrderRepository(ABC):
    """Repository port for Order persistence.

    ABC = zero framework dependencies.
    Infrastructure adapters implement this.
    """

    @abstractmethod
    def save(self, order: Order) -> Order:
        """Persist the order to the database."""
        ...

    @abstractmethod
    def find_by_id(self, order_id: OrderId) -> Optional[Order]:
        """Retrieve an order by its OrderId value object."""
        ...

    @abstractmethod
    def find_by_customer_id(self, customer_id: OrderId) -> List[Order]:
        """Find all orders for a customer by OrderId."""
        ...

    @abstractmethod
    def find_all(self) -> List[Order]:
        """Return all orders."""
        ...

    @abstractmethod
    def count(self) -> int:
        """Return total number of orders."""
        ...

    @abstractmethod
    def exists(self, order_id: OrderId) -> bool:
        """Check if an order exists."""
        ...
